marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/MintTokenEvent.hs

```diff
@@ -256,23 +348,191 @@ mkMintTokenIndexer dbPath = do
+          [sql|
+          SELECT slotNo, blockHeaderHash, txIndexInBlock, txId, policyId, assetName, quantity, redeemerHash, redeemerData
+          FROM minting_policy_events
+          WHERE slotNo <= :slotNo
+          ORDER BY slotNo ASC, txIndexInBlock ASC
+          |]
+
+        parseResult
+          :: (MintTokenBlockEvents -> Maybe MintTokenBlockEvents)
+          -> [Core.Timed C.ChainPoint MintTokenEvent]
+          -> [Core.Timed C.ChainPoint MintTokenBlockEvents]
+        parseResult eventsFilterFunc es =
+          mapMaybe (sequence . over Core.event eventsFilterFunc) $ toTimedMintEvents es
+     in Core.querySyncedOnlySQLiteIndexerWith
+          (\cp -> pure [":slotNo" := C.chainPointToSlotNo cp])
+          (const mintTokenEventQuery)
+          (\(Core.EventsMatchingQuery p) r -> parseResult p r)
+
+data QueryByAssetId event = QueryByAssetId !C.PolicyId !(Maybe C.AssetName) !(Maybe EventType)
+  deriving (Show)
+
+data EventType = MintEventType | BurnEventType
+  deriving (Show)
+
+type instance Core.Result (QueryByAssetId event) = [Core.Timed (Core.Point event) event]
+
+-- TODO Is there a way to do that for reusing the existing instance while enhancing it with
+-- additional features like sorting?
+-- instance
+--   ( MonadIO m
+--   , MonadError (Core.QueryError (Core.EventsMatchingQuery MintTokenBlockEvents)) m
+--   , MonadError (Core.QueryError (QueryByAssetId MintTokenBlockEvents)) m
+--   )
+--   => Core.Queryable m MintTokenBlockEvents (Core.EventsMatchingQuery MintTokenBlockEvents) Core.ListIndexer
+--   where
+--   query point queryPredicate ix = do
+--     pure []
+
+instance
+  ( MonadIO m
+  , MonadError (Core.QueryError (QueryByAssetId MintTokenBlockEvents)) m
+  )
+  => Core.Queryable m MintTokenBlockEvents (QueryByAssetId MintTokenBlockEvents) Core.ListIndexer
+  where
+  query point (QueryByAssetId policyId assetNameM eventType) ix = do
+    -- Filter events based on 'QueryByAssetId' query
+    let queryByAssetIdPredicate = Core.EventsMatchingQuery $ \(MintTokenBlockEvents events) ->
+          let isEventType :: MintTokenEvent -> Bool
+              isEventType e = case eventType of
+                Nothing -> True
+                Just MintEventType -> e ^. mintTokenEventAsset . mintAssetQuantity > 0
+                Just BurnEventType -> e ^. mintTokenEventAsset . mintAssetQuantity < 0
+              isAssetId :: MintTokenEvent -> Bool
+              isAssetId e =
+                case assetNameM of
+                  Nothing ->
+                    e ^. mintTokenEventAsset . mintAssetPolicyId == policyId
+                  Just assetName ->
+                    e ^. mintTokenEventAsset . mintAssetPolicyId == policyId
+                      && e ^. mintTokenEventAsset . mintAssetAssetName == assetName
+           in fmap MintTokenBlockEvents $
+                NonEmpty.nonEmpty $
+                  NonEmpty.filter (\e -> isAssetId e && isEventType e) events
+    -- TODO Find a way to propagate the error.
```

Comment:
Add TODO: Use `mapError` when available

Reply:
No, the Queryable class as a whole, I added a point to talk about it in our weekly technical meeting.